Public/Src/Cache/ContentStore/Library/Service/Grpc/GrpcCopyClient.cs

```diff
@@ -24,28 +26,85 @@ namespace BuildXL.Cache.ContentStore.Service.Grpc
     /// </summary>
     public sealed class GrpcCopyClient : IShutdown<BoolResult>
     {
+        // Link to DistributedContentStoreSettings.MaxConcurrentCopyOperations
+        private const int _maxClientCount = 512;
+
+        private static readonly object lockObject = new object();
+        internal static readonly ConcurrentDictionary<(string, int, bool), GrpcCopyClient> _clientDict = new ConcurrentDictionary<(string, int, bool), GrpcCopyClient>(Environment.ProcessorCount, _maxClientCount);
+        private static Task _backgroundCleaningTask;
+        private static CancellationTokenSource _backgroundCleaningTaskTokenSource;
+
         private readonly Channel _channel;
         private readonly ContentServer.ContentServerClient _client;
         private readonly string _host;
         private readonly int _grpcPort;
         private bool _useCompression;
+        internal DateTime _lastUseTime;

```

Comment:
I would say that this should be stored externally from the client.

Reply:
In the cache's dictionary? That seems more complex than just keeping it tied to the client itself.

---
In reply to: [289207076](https://github.com/microsoft/BuildXL/pull/391#discussion_r289207076) [](ancestors = 289207076)